File: src/domain/nomenclature/flowCategory.ts
/**
 * Catégories d'affichage de l'écoulement ONDE (ADR-006).
 *
 * Quatre catégories de carte — `Ecoulement`, `EcoulementFaible`,
 * `EcoulementNonVisible`, `Assec` — plus deux états qui ne sont PAS des
 * synonymes :
 *
 *  - `NonObserve` : l'observateur s'est déplacé et n'a pas pu observer
 *    (code `4`, « Observation impossible »). C'est un fait de terrain.
 *  - `Inconnu`    : nous ne savons pas lire la donnée — code absent ou non
 *    répertorié. C'est notre ignorance à nous (BR-011).
 *
 * Les fondre ferait passer notre ignorance pour une observation, ce que le
 * produit s'interdit (BR-007). ADR-006 les regroupe à l'affichage ; le domaine,
 * lui, les garde distincts pour que rien ne soit perdu.
 */
export type FlowCategory =
  | "Ecoulement"
  | "EcoulementFaible"
  | "EcoulementNonVisible"
  | "Assec"
  | "NonObserve"
  | "Inconnu";

/**
 * Les six codes constatés en production le 2026-08-01, sur 7 000 observations
 * de `/v1/ecoulement/observations` réparties sur 8 départements :
 *   "1" 1475 · "1a" 2877 · "1f" 1148 · "2" 405 · "3" 1042 · "4" 23
 *
 * Ce sont des chaînes, pas des entiers (C-10).
 */
const BY_CODE: Readonly<Record<string, FlowCategory>> = {
  "1": "Ecoulement",
  "1a": "Ecoulement",
  "1f": "EcoulementFaible",
  "2": "EcoulementNonVisible",
  "3": "Assec",
  "4": "NonObserve",
};

/**
 * `code_ecoulement` vaut `null` sur une part non négligeable des observations —
 * 30 sur 7 000 le 2026-08-01, soit **plus** que le code `4` lui-même. Ce n'est
 * donc pas un cas défensif hypothétique : c'est un cas nominal.
 *
 * Un code non répertorié ne fait pas échouer la lecture non plus. Perdre
 * l'observation serait pire que de nommer notre ignorance.
 */
export function flowCategoryFromOndeCode(code: string | null | undefined): FlowCategory {
  if (code === null || code === undefined) return "Inconnu";
  return BY_CODE[code.trim().toLowerCase()] ?? "Inconnu";
}

File: lib/domain/nomenclature/flow_category.dart
// C-10, reproduit le 2026-09-13 sur 300 observations du departement 41 :
// code_ecoulement est une CHAINE — '1a' 116, '1f' 95, '2' 24, '3' 65. Les
// codes '1' et '4' (releves le 2026-08-01 sur huit departements) n'y
// apparaissent pas : la nomenclature les accepte sans les exiger.
//
// sealed class et non enumeration : Inconnu doit porter la valeur brute
// recue (BR-011), ce qu'une valeur d'enum ne sait pas faire. NonObserve
// (fait de terrain constate) et Inconnu (notre ignorance) restent distincts
// (BR-007) — ADR-006 les regroupe uniquement a l'affichage.

/// Categorie d'ecoulement observee sur une station ONDE.
sealed class FlowCategory {
  const FlowCategory();
}

/// Ecoulement visible — codes `1` et `1a`.
final class Ecoulement extends FlowCategory {
  const Ecoulement();

  @override
  bool operator ==(Object other) => other is Ecoulement;

  @override
  int get hashCode => runtimeType.hashCode;
}

/// Ecoulement faible — code `1f`. Signal precurseur d'un assec a venir.
final class EcoulementFaible extends FlowCategory {
  const EcoulementFaible();

  @override
  bool operator ==(Object other) => other is EcoulementFaible;

  @override
  int get hashCode => runtimeType.hashCode;
}

/// Ecoulement non visible — code `2`. Des flaques persistent, mais plus
/// d'ecoulement.
final class EcoulementNonVisible extends FlowCategory {
  const EcoulementNonVisible();

  @override
  bool operator ==(Object other) => other is EcoulementNonVisible;

  @override
  int get hashCode => runtimeType.hashCode;
}

/// Assec — code `3`. Lit a sec.
final class Assec extends FlowCategory {
  const Assec();

  @override
  bool operator ==(Object other) => other is Assec;

  @override
  int get hashCode => runtimeType.hashCode;
}

/// Non observe — code `4`. Un fait de terrain (observation impossible),
/// jamais confondu avec [Inconnu] qui est notre propre ignorance d'un code
/// (BR-007).
final class NonObserve extends FlowCategory {
  const NonObserve();

  @override
  bool operator ==(Object other) => other is NonObserve;

  @override
  int get hashCode => runtimeType.hashCode;
}

/// Code absent ou non reconnu. Porte la valeur brute [rawCode], telle que
/// recue — jamais normalisee (BR-011) : c'est ce qui distingue cette classe
/// d'une enumeration classique.
final class Inconnu extends FlowCategory {
  const Inconnu(this.rawCode);

  /// La valeur telle que recue de l'API, sans `trim` ni changement de
  /// casse. `null` si aucun code n'a ete transmis.
  final String? rawCode;

  @override
  bool operator ==(Object other) =>
      other is Inconnu && other.rawCode == rawCode;

  @override
  int get hashCode => rawCode.hashCode;

  @override
  String toString() => 'Inconnu($rawCode)';
}

/// Traduit un `code_ecoulement` de l'API ONDE en [FlowCategory]. Ne leve
/// jamais : un code non reconnu devient [Inconnu] plutot que de faire
/// echouer l'appelant. La comparaison se fait apres `trim` et passage en
/// minuscules, mais [Inconnu.rawCode] conserve la valeur recue, non
/// normalisee.
FlowCategory flowCategoryFromCode(String? code) {
  final String? normalized = code?.trim().toLowerCase();

  switch (normalized) {
    case '1':
    case '1a':
      return const Ecoulement();
    case '1f':
      return const EcoulementFaible();
    case '2':
      return const EcoulementNonVisible();
    case '3':
      return const Assec();
    case '4':
      return const NonObserve();
    default:
      return Inconnu(code);
  }
}

/// Libelle affichable d'une [FlowCategory]. `switch` exhaustif : une
/// sous-classe ajoutee sans branche ici est une erreur de compilation
/// (BR-011), pas un oubli silencieux a l'ecran.
String flowCategoryLabel(FlowCategory category) => switch (category) {
  Ecoulement() => 'Écoulement visible',
  EcoulementFaible() => 'Écoulement faible',
  EcoulementNonVisible() => 'Eau stagnante',
  Assec() => 'À sec',
  NonObserve() => 'Observation impossible',
  Inconnu() => 'Non renseigné',
};
